torchvision/transforms/v2/functional/_utils.py

```diff
@@ -1,8 +1,110 @@
+import functools
+import inspect
 from typing import Any
 
 import torch
+from torchvision import datapoints
 from torchvision.datapoints._datapoint import Datapoint
 
 
 def is_simple_tensor(inpt: Any) -> bool:
     return isinstance(inpt, torch.Tensor) and not isinstance(inpt, Datapoint)
+
+
+_KERNEL_REGISTRY = {}
+
+
+def _kernel_wrapper_internal(dispatcher, kernel):
+    dispatcher_params = list(inspect.signature(dispatcher).parameters)[1:]
+    kernel_params = list(inspect.signature(kernel).parameters)[1:]
+
+    needs_args_kwargs_handling = kernel_params != dispatcher_params
+
+    # this avoids converting list -> set at runtime below
+    kernel_params = set(kernel_params)
+
+    @functools.wraps(kernel)
+    def wrapper(inpt, *args, **kwargs):
+        input_type = type(inpt)
+
+        if needs_args_kwargs_handling:
+            # Convert args to kwargs to simplify further processing
+            kwargs.update(dict(zip(dispatcher_params, args)))
+            args = ()
+
+            # drop parameters that are not relevant for the kernel, but have a default value
+            # in the dispatcher
+            for kwarg in kwargs.keys() - kernel_params:
+                del kwargs[kwarg]
+
+            # add parameters that are passed implicitly to the dispatcher as metadata,
+            # but have to be explicit for the kernel
+            for kwarg in input_type.__annotations__.keys() & kernel_params:
+                kwargs[kwarg] = getattr(inpt, kwarg)
+
+        output = kernel(inpt.as_subclass(torch.Tensor), *args, **kwargs)
+
+        if isinstance(inpt, datapoints.BoundingBox) and isinstance(output, tuple):
+            output, spatial_size = output
+            metadata = dict(spatial_size=spatial_size)
+        else:
+            metadata = dict()
+
+        return input_type.wrap_like(inpt, output, **metadata)
+
+    return wrapper
+
+
+def _register_kernel_internal(dispatcher, datapoint_cls, *, wrap_kernel=True):
+    registry = _KERNEL_REGISTRY.setdefault(dispatcher, {})
+    if datapoint_cls in registry:
+        raise TypeError(
+            f"Dispatcher '{dispatcher.__name__}' already has a kernel registered for type '{datapoint_cls.__name__}'."
```

Comment:
My concern here is that a library based on TorchVision registers something for our builtin stuff and thus invalidates our docs if we import the library. Imagine a library does this

```python
@F.register_kernel(F.resize, datapoints.Image)
def lol(...):
    ...
```

I can already see a user of that library coming to our issue tracker with something like

> I see these weird behavior for resize that is not consistent with the docs. Here is my env

And after some back and forth we find that they have `import third_party_library` at the top of his script.

I want to prevent this. Especially since with the subclass check, there is no reason for it. The third-party library can simply do

```python
class CustomImage(datapoints.Image):
    ...

@F.register_kernel(F.resize, CustomImage)
def kernel(...):
    ...
```

Now, if the user deals with plain `datapoints.Image`'s they will always get our behavior.